File: config/config-annotation-processor/src/main/java/ru/tinkoff/kora/config/annotation/processor/ConfigUtils.java
package ru.tinkoff.kora.config.annotation.processor;

import com.squareup.javapoet.ClassName;
import com.squareup.javapoet.TypeName;
import ru.tinkoff.kora.annotation.processor.common.CommonUtils;
import ru.tinkoff.kora.annotation.processor.common.CommonUtils.MappingData;
import ru.tinkoff.kora.annotation.processor.common.ProcessingError;
import ru.tinkoff.kora.common.util.Either;

import javax.annotation.Nullable;
import javax.lang.model.element.*;
import javax.lang.model.type.DeclaredType;
import javax.lang.model.type.ExecutableType;
import javax.lang.model.type.TypeKind;
import javax.lang.model.util.Types;
import java.util.*;
import java.util.stream.Collectors;

public class ConfigUtils {
    public static final Set<TypeName> SUPPORTED_TYPES = Set.of(
        TypeName.INT, TypeName.INT.box(),
        TypeName.LONG, TypeName.LONG.box(),
        TypeName.DOUBLE, TypeName.DOUBLE.box(),
        ClassName.get(String.class)
    );

    public static boolean isSupportedType(TypeName typeName) {
        return SUPPORTED_TYPES.contains(typeName);
    }

    public record ConfigField(String name, TypeName typeName, boolean isNullable, boolean hasDefault, @Nullable MappingData mapping) {}

    public static Either<List<ConfigField>, List<ProcessingError>> parseFields(Types types, TypeElement typeElement) {
        var type = (DeclaredType) typeElement.asType();
        if (typeElement.getKind() == ElementKind.RECORD) {
            return parseRecord(types, type, typeElement);
        } else if (typeElement.getKind() == ElementKind.INTERFACE) {
            return parseInterface(types, type, typeElement);
        } else if (typeElement.getKind() == ElementKind.CLASS) {
            return parseClass(types, type, typeElement);
        } else {
            return Either.right(List.of(new ProcessingError("typeElement should be interface, class or record, got " + typeElement.getKind(), typeElement)));
        }
    }

    private static Either<List<ConfigField>, List<ProcessingError>> parseRecord(Types types, DeclaredType typeMirror, TypeElement te) {
        if (te.getKind() != ElementKind.RECORD) {
            throw new IllegalArgumentException("Method expecting record");
        }
        var fields = new ArrayList<ConfigField>();
        for (var recordComponent : te.getRecordComponents()) {
            var recordComponentType = types.asMemberOf(typeMirror, recordComponent);
            var name = recordComponent.getSimpleName().toString();
            var mapping = CommonUtils.parseMapping(recordComponent).getMapping(ConfigClassNames.configValueExtractor);
            var isNullable = CommonUtils.isNullable(recordComponent) && !recordComponentType.getKind().isPrimitive();
            fields.add(new ConfigUtils.ConfigField(
                name, TypeName.get(recordComponentType), isNullable, false, mapping
            ));
        }
        return Either.left(fields);
    }

    private static Either<List<ConfigField>, List<ProcessingError>> parseInterface(Types types, DeclaredType typeMirror, TypeElement te) {
        if (te.getKind() != ElementKind.INTERFACE) {
            throw new IllegalArgumentException("Method expecting interface");
        }
        var seen = new HashSet<String>();
        var errors = new ArrayList<ProcessingError>();
        var fields = new ArrayList<ConfigField>();

        parseInterface(types, typeMirror, te, fields, errors, seen);
        if (errors.isEmpty()) {
            return Either.left(fields);
        } else {
            return Either.right(errors);
        }
    }

    private static void parseInterface(Types types, DeclaredType typeMirror, TypeElement te, List<ConfigField> fields, List<ProcessingError> errors, Set<String> seen) {
        if (te.getKind() != ElementKind.INTERFACE) {
            throw new IllegalArgumentException("Method expecting interface");
        }
        for (var enclosedElement : te.getEnclosedElements()) {
            if (enclosedElement.getKind() != ElementKind.METHOD || enclosedElement.getModifiers().contains(Modifier.STATIC) || enclosedElement.getModifiers().contains(Modifier.PRIVATE)) {
                continue;
            }
            var method = (ExecutableElement) enclosedElement;
            if (!method.getParameters().isEmpty()) {
                if (method.getModifiers().contains(Modifier.DEFAULT)) {
                    continue;
                } else {
                    errors.add(new ProcessingError("Config has non default method with arguments", method));
                }
            }
            if (method.getReturnType().getKind() == TypeKind.VOID) {
                if (method.getModifiers().contains(Modifier.DEFAULT)) {
                    continue;
                }
                errors.add(new ProcessingError("Config has non default method returning void", method));
            }
            if (!method.getTypeParameters().isEmpty()) {
                errors.add(new ProcessingError("Config has method with type parameters", method));
            }
            var methodType = (ExecutableType) types.asMemberOf(typeMirror, method);
            var name = method.getSimpleName().toString();
            if (seen.add(name)) {
                var isNullable = CommonUtils.isNullable(method) && !methodType.getReturnType().getKind().isPrimitive();
                var mapping = CommonUtils.parseMapping(method).getMapping(ConfigClassNames.configValueExtractor);
                fields.add(new ConfigUtils.ConfigField(
                    name, TypeName.get(methodType.getReturnType()), isNullable, method.getModifiers().contains(Modifier.DEFAULT), mapping
                ));
            }
            for (var superinterface : te.getInterfaces()) {
                var superinterfaceElement = (TypeElement) types.asElement(superinterface);
                parseInterface(types, (DeclaredType) superinterface, superinterfaceElement, fields, errors, seen);
            }
        }
    }

    private static Either<List<ConfigField>, List<ProcessingError>> parseClass(Types types, DeclaredType typeMirror, TypeElement te) {
        var errors = new ArrayList<ProcessingError>();
        if (te.getKind() != ElementKind.CLASS) {
            throw new IllegalArgumentException("Method expecting record");
        }
        if (te.getModifiers().contains(Modifier.ABSTRACT)) {
            errors.add(new ProcessingError("Config annotated class can't be abstract", te));
            return Either.right(errors);
        }
        ExecutableElement equals = null;
        ExecutableElement hashCode = null;
        class FieldAndAccessors {
            VariableElement field;
            ExecutableElement getter;
            ExecutableElement setter;
        }
        var fieldsWithAccessors = new HashMap<String, FieldAndAccessors>();
        for (var enclosedElement : te.getEnclosedElements()) {
            var name = enclosedElement.getSimpleName().toString();
            if (enclosedElement.getKind() == ElementKind.FIELD) {
                fieldsWithAccessors.computeIfAbsent(name, n -> new FieldAndAccessors()).field = (VariableElement) enclosedElement;
            }
            if (enclosedElement.getKind() == ElementKind.METHOD) {
                if (enclosedElement.getModifiers().contains(Modifier.PRIVATE)) {
                    continue;
                }
                if (enclosedElement.getModifiers().contains(Modifier.STATIC)) {
                    continue;
                }
                var method = (ExecutableElement) enclosedElement;
                if (name.equals("equals") && method.getParameters().size() == 1) {
                    equals = method;
                } else if (name.equals("hashCode") && method.getParameters().isEmpty()) {
                    hashCode = method;
                } else {
                    if (name.startsWith("get")) {
                        fieldsWithAccessors.computeIfAbsent(CommonUtils.decapitalize(name.substring(3)), n -> new FieldAndAccessors()).getter = method;
                    } else if (name.startsWith("set")) {
                        fieldsWithAccessors.computeIfAbsent(CommonUtils.decapitalize(name.substring(3)), n -> new FieldAndAccessors()).setter = method;
                    }
                }
            }
        }
        if (equals == null || hashCode == null) {
            errors.add(new ProcessingError("Config annotated class must override equals and hashCode methods", te));
            return Either.right(errors);
        }
        var constructors = CommonUtils.findConstructors(te, m -> m.contains(Modifier.PUBLIC));
        var emptyConstructor = constructors.stream().filter(e -> e.getParameters().isEmpty()).findFirst().orElse(null);
        var nonEmptyConstructor = constructors.stream().filter(e -> !e.getParameters().isEmpty()).findFirst().orElse(null);
        var constructorParams = nonEmptyConstructor == null ? Set.of() : nonEmptyConstructor.getParameters().stream().map(VariableElement::getSimpleName).map(Objects::toString).collect(Collectors.toSet());

        var seen = new HashSet<String>();
        var fields = new ArrayList<ConfigField>();
        for (var fieldWithAccessors : fieldsWithAccessors.entrySet()) {
            var name = fieldWithAccessors.getKey();
            var value = fieldWithAccessors.getValue();
            if (value.getter == null) {
                continue;
            }
            if (value.setter == null && !constructorParams.contains(value.field.getSimpleName().toString())) {
                continue;
            }
            var fieldType = types.asMemberOf(typeMirror, value.field);
            if (seen.add(name)) {
                var isNullable = CommonUtils.isNullable(value.field) && !fieldType.getKind().isPrimitive();
                var mapping = CommonUtils.parseMapping(value.field).getMapping(ConfigClassNames.configValueExtractor);
                var hasDefault = emptyConstructor != null || !constructorParams.contains(value.field.getSimpleName().toString());
                fields.add(new ConfigUtils.ConfigField(
                    name, TypeName.get(fieldType), isNullable, hasDefault, mapping
                ));
            }
        }
        return Either.left(fields);
    }

}

File: config/config-symbol-processor/src/main/kotlin/ru/tinkoff/kora/config/ksp/ConfigUtils.kt
package ru.tinkoff.kora.config.ksp

import com.google.devtools.ksp.getConstructors
import com.google.devtools.ksp.getDeclaredFunctions
import com.google.devtools.ksp.processing.Resolver
import com.google.devtools.ksp.symbol.*
import com.squareup.kotlinpoet.TypeName
import com.squareup.kotlinpoet.ksp.toTypeName
import ru.tinkoff.kora.common.util.Either
import ru.tinkoff.kora.ksp.common.JavaUtils.isRecord
import ru.tinkoff.kora.ksp.common.JavaUtils.recordComponents
import ru.tinkoff.kora.ksp.common.MappingData
import ru.tinkoff.kora.ksp.common.exception.ProcessingError
import ru.tinkoff.kora.ksp.common.parseMappingData

object ConfigUtils {
    fun parseFields(resolver: Resolver, element: KSClassDeclaration): Either<List<ConfigField>, List<ProcessingError>> {
        val errors = arrayListOf<ProcessingError>()
        val seen = hashSetOf<String>()
        val fields = arrayListOf<ConfigField>()

        fun parseRecordFields(type: KSType, typeDecl: KSClassDeclaration) {
            require(typeDecl.isRecord()) { "should be record" }
            for (recordComponent in typeDecl.recordComponents()) {
                val recordComponentType = recordComponent.asMemberOf(type)
                val name = recordComponent.simpleName.asString()
                if (seen.add(name)) {
                    val isNullable = recordComponentType.isMarkedNullable
                    val mapping = recordComponent.parseMappingData().getMapping(ConfigClassNames.configValueExtractor)
                    fields.add(
                        ConfigField(
                            name, recordComponentType.toTypeName(), isNullable, false, mapping
                        )
                    )
                }
            }
        }

        fun parsePojoFields(type: KSType, typeDecl: KSClassDeclaration) {
            require(typeDecl.classKind == ClassKind.CLASS) { "Method expecting class" }
            if (typeDecl.modifiers.contains(Modifier.ABSTRACT)) {
                errors.add(ProcessingError("Config annotated class can't be abstract", typeDecl))
                return
            }
            var equals: KSFunctionDeclaration? = null
            var hashCode: KSFunctionDeclaration? = null
            class FieldAndAccessors(var property: KSPropertyDeclaration? = null, var getter: KSFunctionDeclaration? = null, var setter: KSFunctionDeclaration? = null)
            val propertyMap = hashMapOf<String, FieldAndAccessors>()
            for (function in typeDecl.getDeclaredFunctions()) {
                val name = function.simpleName.asString()
                if (function.modifiers.contains(Modifier.PRIVATE)) {
                    continue
                }
                if (name == "equals" && function.parameters.size == 1) {
                    equals = function
                } else if (name == "hashCode" && function.parameters.isEmpty()) {
                    hashCode = function
                } else {
                    if (name.startsWith("get")) {
                        val propertyName = name.substring(3).replaceFirstChar { it.lowercaseChar() }
                        propertyMap.computeIfAbsent(propertyName) {FieldAndAccessors()}.getter = function
                    }
                    if (name.startsWith("set")) {
                        val propertyName = name.substring(3).replaceFirstChar { it.lowercaseChar() }
                        propertyMap.computeIfAbsent(propertyName) {FieldAndAccessors()}.setter = function
                    }
                }
            }
            for (property in typeDecl.getAllProperties()) {
                propertyMap.computeIfAbsent(property.simpleName.asString()) {FieldAndAccessors()}.property = property
            }
            val properties = propertyMap.values.asSequence().filter { it.setter != null && it.getter != null && it.property != null }.map { it.property!! }.toList()
            if (equals == null || hashCode == null) {
                errors.add(ProcessingError("Config annotated class must override equals and hashCode methods", typeDecl))
                return
            }
            properties.forEach {
                val fieldType = it.asMemberOf(type)
                if (seen.add(it.simpleName.asString())) {
                    val isNullable = fieldType.isMarkedNullable
                    val mapping = it.parseMappingData().getMapping(ConfigClassNames.configValueExtractor)
                    fields.add(ConfigField(it.simpleName.asString(), fieldType.toTypeName(), isNullable, true, mapping))
                }
            }
        }

        fun parseDataClassFields(type: KSType, typeDecl: KSClassDeclaration) {
            require(typeDecl.classKind == ClassKind.CLASS) { "Method expecting class" }
            require(typeDecl.modifiers.contains(Modifier.DATA))
            if (typeDecl.modifiers.contains(Modifier.ABSTRACT)) {
                errors.add(ProcessingError("Config annotated class can't be abstract", typeDecl))
                return
            }
            val primaryConstructor = typeDecl.primaryConstructor!!
            for (parameter in primaryConstructor.parameters) {
                val name = parameter.name!!.asString()
                val fieldType = parameter.type.resolve()
                val isNullable = fieldType.isMarkedNullable
                val mapping = parameter.parseMappingData().getMapping(ConfigClassNames.configValueExtractor)
                fields.add(ConfigField(name, fieldType.toTypeName(), isNullable, false, mapping))
            }
        }

        fun parseInterfaceFields(type: KSType, typeDecl: KSClassDeclaration) {
            require(typeDecl.classKind == ClassKind.INTERFACE) { "Method expecting interface" }
            for (function in typeDecl.getAllFunctions()) {
                when (function.simpleName.asString()) {
                    "equals" -> {
                        if (function.parameters.size == 1) {
                            continue
                        }
                    }

                    "hashCode" -> {
                        if (function.parameters.isEmpty()) {
                            continue
                        }
                    }

                    "toString" -> {
                        if (function.parameters.isEmpty()) {
                            continue
                        }
                    }

                }
                if (function.parameters.isNotEmpty()) {
                    if (!function.isAbstract) {
                        // todo are default java functions abstract?
                        continue
                    } else {
                        errors.add(ProcessingError("Config has non default method with arguments: ${function.simpleName.asString()}", function))
                        return
                    }
                }
                if (function.returnType == resolver.builtIns.unitType) {
                    if (!function.isAbstract) {
                        continue
                    }
                    errors.add(ProcessingError("Config has non default method returning void", function))
                    return
                }
                if (function.typeParameters.isNotEmpty()) {
                    errors.add(ProcessingError("Config has method with type parameters", function))
                    return
                }
                val functionType = function.asMemberOf(type)
                val name = function.simpleName.asString()
                if (seen.add(name)) {
                    val isNullable = functionType.returnType!!.isMarkedNullable
                    val mapping = function.parseMappingData().getMapping(ConfigClassNames.configValueExtractor)
                    fields.add(
                        ConfigField(
                            name, functionType.returnType!!.toTypeName(), isNullable, !function.isAbstract, mapping
                        )
                    )
                }
            }
        }

        val type = element.asType(listOf())
        if (element.isRecord()) {
            parseRecordFields(type, element)
        } else if (element.classKind == ClassKind.INTERFACE) {
            parseInterfaceFields(type, element)
        } else if (element.classKind == ClassKind.CLASS && element.modifiers.contains(Modifier.DATA)) {
            parseDataClassFields(type, element)
        } else if (element.classKind == ClassKind.CLASS && element.getConstructors().any { it.modifiers.contains(Modifier.PUBLIC) && it.parameters.isEmpty() }) {
            parsePojoFields(type, element)
        } else {
            return Either.right(
                listOf(
                    ProcessingError(
                        "typeElement should be interface, data class or java record, got " + element.classKind,
                        element
                    )
                )
            )
        }
        return if (errors.isEmpty()) {
            Either.left(fields)
        } else {
            Either.right(errors)
        }
    }

    data class ConfigField(val name: String, val typeName: TypeName, val isNullable: Boolean, val hasDefault: Boolean, val mapping: MappingData?)
}
